### DISCUSSION TITLE
Downloadable Products

### DISCUSSION THREAD

**Discussion by feelfetch** [medusajs/medusa#2228] [Q&A] [answered]
This is such a needed feature!

**Comment by olivermrbl** [comment]
@feelfetch Thanks for reporting it in. What is your use case? 

We've helped numerous merchants open businesses with downloadable products, so I assume we could also assist you.

Usually, the desired customer flow is attaching the downloadable product in the order confirmation email, but you might have additional requirements.

**Reply by feelfetch** [reply]
That would be awesome! Yes, attach a download link to order confirmation page and email. That is more than enough for my use case.

**Reply by slavanga** [reply]
@olivermrbl I'm also very interested in how to achieve downloadable products like you described with Medusa.

**Comment by srindom** [comment, marked as answer]
Where would you want to store the downloadable file?

I could see a few options for how this could work with what we have currently and then how it could be done in the future.

For the current setup you could do this:

- In your Medusa project/plugin, create a custom entity `src/models/downloadable-asset`
  - Model will have `id`, `variant_id`, `file_url` (maybe also `title`, etc, etc)
- Create a custom Service that handles the business logic - it will have:
  - `create` - takes variant and file_url, ensures uniqueness on variant id etc.
  - `update` - can be used to change the file_url
  - `delete` - use to an asset from a product
  - `listForOrder` - checks what downloadable assets are available for an order by going through order.items.$.variant and checking for `downloadable-assets`
  - (`createSignedUrl` - maybe not necessary, but could be relevant if you want to create signed URLs that are only valid for a given time-period)
- Create API endpoints to manage assets
- Modify the product detail page in medusajs/admin to allow adding downloadable assets to variants (use the custom API endpoints you created).
- Note you can use the FileService to upload, store and manage your files (see `medusa-file-*` plugins in `/packages`).
- Create a subscriber that listens to the `order.placed` event; the handler for the event should do:
  - `downloadableAssetService.listForOrder(orderId)`; if there are any downloadable assets issue send an email to the customer with the `file_url`s. Note you can use a notification plugin to handle the email send (e.g., `medusa-plugin-sendgrid`).

^ This could be made more advanced depending on your needs, but think this is a good, simple starting point

Btw, this would be such a cool plugin to create!

--
In the future, I could imagine this functionality being supported out of the box, but it is not something we have on the roadmap at the moment.

**Reply by slavanga** [reply]
@srindom thank you for the detailed answer. This will help for my further evaluations. Would definitely be nice to see this as a feature one day.
